src/modules/dotnet.rs

```diff
@@ -46,13 +49,89 @@ pub fn module<'a>(context: &'a Context) -> Option<Module<'a>> {
         get_version_from_cli()?
     };
 
-    module.set_style(config.style);
-    module.create_segment("symbol", &config.symbol);
-    module.create_segment("version", &config.version.with_value(&version.0));
+    let current_tfm = find_current_tfm(&dotnet_files);
+
+    let formatter = if let Ok(formatter) = StringFormatter::new(config.format) {
+        formatter
+            .map_style(|variable| match variable {
+                "style" => Some(config.style.to_string()),
+                _ => None,
+            })
+            .map(|variable| match variable {
+                "symbol" => Some(config.symbol.to_string()),
+                "version" => Some(version.0.to_string()),
```

Comment:
I'm not super familiar with the `StringFormatter`, but is it possible to just move `version` and `current_tfm` rather than copying them?

Reply:
Yes, it's currently possible to move them in `feat/format-string`.

```
                "version" => Some(version.0),
```

I'm going to review all these changes once the design of `StringFormatter` is mature.

Reply:
Thanks :) I think I have this updated correctly.

Reply:
These are still copying, actually. Your closure returns `Option<&str>` and the `Into<String>` on map will happily clone them.

I cloned your branch and tried to make this work, but unfortunately `map` accepts a `Fn` rather than a `FnOnce`, so you can't move values into that closure anyway. Plus, having an `FnOnce` doesn't make too much sense with the multiple invocations required with different variables. Solving this would probably require changing the `StringFormatter` to take a `HashMap`, or similar, rather than using this multiply-invoked function.

Anyway, I don't think this needs to be solved in your branch.

You can't solve the copying, but I think you can still neaten up a bit:

```
.map(|variable| match variable {
    "symbol" => Some(config.symbol),
    "version" => Some(&version.0),
    "tfm" => current_tfm.as_deref(),
    _ => None,
})
```

Reply:
I do have solved the copying with `Into<Cow<'a, str>>` in another PR (#1148). This PR has some API changes and I'm going to leave it a while for review. Even if the design changes are rejected, I'll open a PR that reduces unnecessary copying wherever  possible, so do not worry.

Reply:
@heyrict - Yes, the `Cow` with the lifetime of the `StringFormatter` looks like a good solution 👍

Reply:
Sorry I am new to Rust. How is the best way to determine if `dotnet_files` is thread safe?

Reply:
If you don't know, just make the changes and have the compiler determine it. A variable(struct, enum, etc.) can be safely shared between threads only when all its members can be safely shared between threads, which is indicated by trait Sync.

Reply:
Yeah, just try it. It's a `Vec` of a simple data type, and you're already consuming it by shared reference, so I'd expect it to be fine.

Reply:
Sorry I think this one is outside my knowledge. I tried to make the update but I had to flow the change from `String` to `&str` all the way through the various functions.

Reply:
It seems that you have to use two `map` as one cannot return two different types in the same function.

Should we change the API to return a `Cow<str>` or `Box<dyn Into<Cow<str>>>`? Though I'm personally happy with the current implementation.

Reply:
That would be a question for someone else sorry I am starting to understand what you are talking about but not quite there yet

Reply:
Oh, sorry. I should have given a better explanation on what happened.

Here is the type of `StringFormatter::map`:

```rust
pub fn map<T: Into<String>>(mut self, mapper: impl Fn(&str) -> Option<T> + Sync) -> Self;
```

This method accepts a callback function `Fn(&str) -> Option<T> + Sync` that converts the variable name (of type `&str`) to a type `Option<impl Into<String>> + Sync`.

The type `Option<impl Into<String>> + Sync` is a `Option<impl Into<String>>` that can be shared safely between threads. and `Into<String>` means the type can be converted to string, e.g. `&str`, or `String` itself.

While this method is flexible and can accept any type of `T` that implements `Into<String>`, there is only one type `T` and cannot accept both `&str` and `String` as a return type, though they are both valid `impl Into<String>`.

An alternative in rust is `dyn Into<String>` that is a dynamic type that accepts both `String` and `&str`, but since it is dynamic, rust compiler does not know its size (it does not implements `Sized` trait). So it should be allocated in the heap, as the size of pointer is fixed for them. That's why `Box<dyn Into<String>>` works.

Another alternative is to convert `&str` to `String` manually.